_drafts/unittests-cpputest.md

```diff
@@ -0,0 +1,1251 @@
+
+[AddressSanitizer](https://github.com/google/sanitizers/wiki/AddressSanitizer)
+
+## C/C++ Unit Tests: Common Issues
+
+Writing unit tests in C isn't as simple as writing tests in some languages. Here
+are some common errors and mistakes that everyone runs into and possible
+solutions.
+
+{:.no_toc}
+
+### Linker Error: Symbol not found
+
+```
+Linking build/kv_store/kv_store_tests
+Undefined symbols for architecture x86_64:
+  "_analytics_inc", referenced from:
+      _kv_store_write in libkv_store.a(kv_store.o)
+      _kv_store_read in libkv_store.a(kv_store.o)
+      _kv_store_delete in libkv_store.a(kv_store.o)
+ld: symbol(s) not found for architecture x86_64
+```
+
+This error is generated by the linker to tell the user that there are undefined
+symbols. In the example above, the function `analytics_inc` is called from three
+different functions, but isn't defined anywhere.
+
+The possible solutions to the issue are:
+
+- Create a fake, stub, or mock file which implements this function and add it
+  either to the headers or compile it into the library
+- Define the function within the unit test file itself, probably at the top of
+  the file. If it is a C function being called by C code, place it within the
+  `extern "C" {}` section.
+- Compile out the call points using a define for unit tests. e.g.
+  `#if !INSIDE_UNITTESTS`
+
+{:.no_toc}
+
+### Linker Error: Duplicate symbol
+
+```
+Linking build/kv_store/kv_store_tests
+duplicate symbol '_mutex_create' in:
+    build/kv_store/objs/complex/tests/src/test_kv_store.o
+    build/kv_store/lib/libkv_store.a(fake_mutex.o)
+ld: 1 duplicate symbol for architecture x86_64
+```
+
+This error is generated by the linker when more than one implementation of a
+function is found and used. This is usually because more than one of the
+following were included in the unit test: real implementation, fake, stub, or
+mock.
+
+In the example above, I had included a `fake_mutex.c` file _and_ included the
+`stub_mutex.h` header, which caused a duplicate `mutex_create` symbol. The
+solution would be to remove one or the other.
+
+{:.no_toc}
+
+### State carrying over between tests
+
+If there is a fake or module which contains static or global state, and it is
+being used across multiple tests in a single file, then that state ideally
+should be cleared out. This is usually done by:
+
+- Defining a `fake_<module>_reset` if the module is a fake.
+- Defining a `<module>_deinit` if the module is a real one that is used in
+  production code. Make sure to compile this function out of the production code
+  by using `#if !INSIDE_UNITTESTS`, or ensuring the linker removes it from the
+  final binary. Code space is precious!
+
+## Final Thoughts
+
+Unit testing was something that a co-worker of mine of mine suggested to me 4
+years ago when writing a complicated raw flash storage to filesystem migration
+for our firmware. After having spent a month doing the **1.** Write Code, **2.**
```

Comment:
maybe: did you intend for this to be a numbered list?

Reply:
I actually wanted it to be an inline list, since our CSS really bloats the lists.